lib/ar_collection_setter.rb

```diff
@@ -1,19 +1,5 @@
 AR_COLLECTION_SETTER = ->(input:, binding:, **) do
-   getter = binding[:getter]
-   collection = getter.nil? ? send(binding.getter) : getter.evaluate(self)
-
-  # Hard-delete records that are being replaced
-  # Any further dependent records must be handled with foreign key constraints
-  if collection.is_a?(ActiveRecord::Associations::CollectionProxy)
-    if collection.respond_to?(:loaded?) && !collection.loaded?
-      collection.delete_all :delete_all
-    else
-      collection.group_by(&:class).each do |klass, items|
-        klass.where(id: items.map(&:id)).delete_all
```

Comment:
I believe this was the problem, because if the tags association reaches this line, then `klass` is just Tag and it deletes those tags (from ALL exercises). In contrast, the `delete_all` above correctly handles has_many through associations by just deleting the join record instead.